File: python-legacy/app/services/budget.py
"""Credit budget tracker from response headers."""

from __future__ import annotations


class BudgetTracker:
    """Tracks API credit usage from response headers."""

    def __init__(self, low_warning: int = 50, critical_stop: int = 10) -> None:
        self.remaining: int | None = None
        self.used: int | None = None
        self.low_warning = low_warning
        self.critical_stop = critical_stop

    def update(self, remaining: int | None, used: int | None) -> None:
        if remaining is not None:
            if self.remaining is None or remaining <= self.remaining:
                self.remaining = remaining
        if used is not None:
            if self.used is None or used >= self.used:
                self.used = used

    @property
    def is_low(self) -> bool:
        if self.remaining is None:
            return False
        return self.remaining <= self.low_warning

    @property
    def is_critical(self) -> bool:
        if self.remaining is None:
            return False
        return self.remaining <= self.critical_stop

    @property
    def can_fetch_odds(self) -> bool:
        """Whether we have enough credits for an odds fetch. Blocks when low."""
        if self.remaining is None:
            return True  # Unknown budget, allow
        return not self.is_low

    @property
    def can_fetch_scores(self) -> bool:
        """Scores allowed when low, blocked only when critical."""
        if self.remaining is None:
            return True
        return not self.is_critical

    @property
    def can_fetch_props(self) -> bool:
        """Props require multiple per-event calls — use a higher threshold."""
        if self.remaining is None:
            return True
        return self.remaining > self.critical_stop * 3

    @property
    def status_text(self) -> str:
        if self.remaining is None:
            return "Credits: --"
        return f"Credits: {self.remaining}"

    @property
    def warning_text(self) -> str:
        if self.is_critical:
            return "CREDITS CRITICAL - Pausing all API calls"
        if self.is_low:
            return "Credits low - Scores only mode"
        return ""

File: src-tauri/src/store/budget.rs
use crate::models::BudgetState;

#[derive(Debug, Clone)]
pub struct BudgetTracker {
    remaining: Option<u32>,
    used: Option<u32>,
    pub low_warning: u32,
    pub critical_stop: u32,
}

impl BudgetTracker {
    pub fn new(low_warning: u32, critical_stop: u32) -> Self {
        Self {
            remaining: None,
            used: None,
            low_warning,
            critical_stop,
        }
    }

    pub fn remaining(&self) -> Option<u32> {
        self.remaining
    }

    pub fn used(&self) -> Option<u32> {
        self.used
    }

    /// Monotonic update. `remaining` only moves down (ignore stale higher
    /// values from out-of-order responses). `used` only moves up.
    ///
    /// Call [`reset`] if credits genuinely jump (API key rotation, refill).
    pub fn update(&mut self, remaining: Option<u32>, used: Option<u32>) {
        if let Some(r) = remaining {
            if self.remaining.map_or(true, |cur| r <= cur) {
                self.remaining = Some(r);
            }
        }
        if let Some(u) = used {
            if self.used.map_or(true, |cur| u >= cur) {
                self.used = Some(u);
            }
        }
    }

    pub fn reset(&mut self) {
        self.remaining = None;
        self.used = None;
    }

    pub fn is_low(&self) -> bool {
        matches!(self.remaining, Some(r) if r <= self.low_warning)
    }

    pub fn is_critical(&self) -> bool {
        matches!(self.remaining, Some(r) if r <= self.critical_stop)
    }

    pub fn can_fetch_odds(&self) -> bool {
        match self.remaining {
            None => true,
            Some(_) => !self.is_low(),
        }
    }

    pub fn can_fetch_scores(&self) -> bool {
        match self.remaining {
            None => true,
            Some(_) => !self.is_critical(),
        }
    }

    pub fn can_fetch_props(&self) -> bool {
        match self.remaining {
            None => true,
            Some(r) => r > self.critical_stop * 3,
        }
    }

    pub fn status_text(&self) -> String {
        match self.remaining {
            None => "Credits: --".to_string(),
            Some(r) => format!("Credits: {r}"),
        }
    }

    pub fn warning_text(&self) -> String {
        if self.is_critical() {
            "CREDITS CRITICAL - Pausing all API calls".to_string()
        } else if self.is_low() {
            "Credits low - Scores only mode".to_string()
        } else {
            String::new()
        }
    }

    pub fn snapshot(&self) -> BudgetState {
        BudgetState {
            remaining: self.remaining,
            used: self.used,
            is_low: self.is_low(),
            is_critical: self.is_critical(),
            status_text: self.status_text(),
            warning_text: self.warning_text(),
        }
    }
}
